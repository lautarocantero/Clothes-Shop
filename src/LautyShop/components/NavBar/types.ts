

export type AppBarColor = 'default' | 'primary' | 'secondary' | 'transparent' | 'inherit';

export type NavBarLink = {
  linkText: string,
  linkUrl: string, 
}