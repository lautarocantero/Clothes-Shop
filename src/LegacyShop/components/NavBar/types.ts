export type AppBarColor = 'default' | 'primary' | 'secondary' | 'transparent' | 'inherit';

export type NavBarLink = {
  linkText: string,
  linkUrl: string, 
}

export interface NavBarProps {
  links: NavBarLink[];
  color?: AppBarColor;
}
