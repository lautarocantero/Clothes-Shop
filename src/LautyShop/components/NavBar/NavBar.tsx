// import { MenuOutlined } from "@mui/icons-material";
// import PersonIcon from '@mui/icons-material/Person';
// import ShoppingCartIcon from '@mui/icons-material/ShoppingCart';
import {
  AppBar,
  GridLegacy as Grid,
  IconButton,
  Link,
  Menu,
  MenuItem,
  Toolbar,
  Typography
} from "@mui/material";
import PropTypes from 'prop-types';
import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link as LinkReactRouter, useNavigate } from "react-router-dom";
import { startLogout, type RootState } from "../../../store/auth";
import type { AppBarColor, NavBarLink } from "./types";

interface NavBarProps {
  links: NavBarLink[];
  color?: AppBarColor;
}

const NavBar = ({ links = [], color = 'primary' }: NavBarProps) => {

  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const [mobileMenuOpen, setMobileMenuOpen] = useState(false);
  const handleNavigate = useNavigate();

  const { name, status } = useSelector((state: RootState) => state?.auth);

  const dispatch = useDispatch();

  const open = Boolean(anchorEl);

  const handleMouseEnter = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMouseLeave = () => {
    setAnchorEl(null);
  };

  const onLogout = () => {
    dispatch(startLogout() as any);
    handleNavigate('/auth/login');
  };

  const onRegister = () => {
    dispatch(startLogout() as any);
    handleNavigate('/auth/register');
  }

  const onHandleShoppingCart = () => {
    handleNavigate('/mi-carrito')
  }

  const DesktopMenuLinks = ({ links }: { links: NavBarLink[] }) => {
    return (
      <>
        {links?.map((linkObject: NavBarLink) => {
          const linkTo = linkObject?.linkUrl?.startsWith('/')
            ? linkObject.linkUrl
            : `/${linkObject.linkUrl}`;

          return (
            <Grid
              item
              key={crypto.randomUUID()}
              sx={{ margin: { xs: '10px', sm: '0 10px' } }}
            >
              <Link
                component={LinkReactRouter}
                to={linkTo}
                color="inherit"
                sx={{
                  fontSize: theme => theme?.typography?.body1?.fontSize,
                  textDecoration: 'none',
                  width: '100%',
                  fontWeight: 200,
                  '&:hover': {
                    color: theme => theme.custom.white,
                    textDecoration: 'underline',
                  },
                }}
              >
                {linkObject?.linkText}
              </Link>
            </Grid>
          );
        })}
      </>
    );
  };

  const DesktopMenu = () => {
    return (
      <>
        <Grid
          item
          xs={12}
          sm="auto"
          container
          justifyContent="center"
          alignItems="center"
          sx={{
            flexGrow: 1,
            flexDirection: { xs: 'column', sm: 'row' },
            display: { xs: 'none', sm: 'flex' },
          }}
        >
          <DesktopMenuLinks links={links} />
        </Grid>
        <Grid item sx={{ display: { xs: 'none', sm: 'flex' } }}>
          <IconButton
            sx={{
              color: theme => theme?.custom?.white,
              fontSize: theme => theme?.typography?.h4?.fontSize,
            }}
            onClick={ () => {handleNavigate('/carrito') }}
          >
            {/* <ShoppingCartIcon /> */}
          </IconButton>
          <IconButton
            onMouseEnter={handleMouseEnter}
            sx={{
              color: theme => theme?.custom?.white,
              fontSize: theme => theme?.typography?.h4?.fontSize,
            }}
          >
            {/* <PersonIcon /> */}
            <Typography>Cuenta</Typography>
          </IconButton>

          <Menu
            open={open}
            onClose={handleMouseLeave}
            MenuListProps={{
              onMouseLeave: handleMouseLeave,
            }}
            sx={{ marginTop: '2.5em'}}
            slotProps={{
          paper: {
            sx: {
              backgroundColor: theme => theme.palette.primary.main,
              color: theme => theme.custom.white,
              borderRadius: 0,
              boxShadow: 'none',
            },
          },
        }}
            anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
            transformOrigin={{ vertical: 'top', horizontal: 'right' }}
          
          >
            <MenuItem>Hola {name || 'Usuario'}!</MenuItem>
            {
              status === 'authenticated' ? (
                <>
                  <MenuItem onClick={onHandleShoppingCart}>Mi carrito</MenuItem>
                  <MenuItem onClick={onLogout}>Cerrar sesión</MenuItem>
                </>
              ) : (
                <>
                  <MenuItem onClick={onRegister} >Registro</MenuItem>
                  <MenuItem onClick={() => handleNavigate('/auth/login')}>Abrir sesión</MenuItem>
                </>
              )
            }
          </Menu>
        </Grid>
      </>
    );
  };

  const MobileMenu = () => {
    return (
      <Menu
        anchorPosition={{ top: 50, left: 0 }}
        anchorReference="anchorPosition"
        anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
        open={mobileMenuOpen}
        onClose={() => setMobileMenuOpen(false)}
        sx={{ display: { xs: 'block', sm: 'none' } }}
        slotProps={{
          paper: {
            sx: {
              width: '80%',
              backgroundColor: theme => theme.palette.primary.main,
              color: theme => theme.custom.white,
              borderRadius: 0,
              boxShadow: 'none',
            },
          },
        }}
      >
        {links?.map((linkObject: NavBarLink) => {
          const linkTo = linkObject?.linkUrl?.startsWith('/')
            ? linkObject.linkUrl
            : `/${linkObject.linkUrl}`;

          return (
            <MenuItem
              key={crypto.randomUUID()}
              component={LinkReactRouter}
              to={linkTo}
              onClick={() => setMobileMenuOpen(false)}
              sx={{ fontSize: theme => theme?.typography?.h5?.fontSize }}
            >
              {linkObject?.linkText}
            </MenuItem>
          );
        })}
        <MenuItem sx={{ fontSize: theme => theme?.typography?.h5?.fontSize }}>
          Hola {name || 'Usuario'}!
        </MenuItem>
        <MenuItem sx={{ fontSize: theme => theme?.typography?.h5?.fontSize }}>
          Perfil
        </MenuItem>
        <MenuItem onClick={onLogout}>Cerrar sesión</MenuItem>
      </Menu>
    );
  };

  return (
    <>
      <AppBar color={color}>
        <Toolbar>
          <Grid
            container
            alignItems="center"
            justifyContent="space-between"
            sx={{
              flexDirection: { xs: 'column', sm: 'row' },
              alignContent: { xs: 'end', sm: 'center' },
            }}
          >
            <Grid item>
              <IconButton
                aria-label="Abrir menú de usuario"
                sx={{ display: { xs: 'block', sm: 'none' }, color: theme => theme?.custom?.white }}
                onClick={() => setMobileMenuOpen(true)}
              >
                {/* <MenuOutlined /> */}
              </IconButton>
            </Grid>
            <DesktopMenu />
          </Grid>
        </Toolbar>
      </AppBar>
      <MobileMenu />
    </>
  );
};

NavBar.propTypes = {
  links: PropTypes.array.isRequired,
  color: PropTypes.string,
};

export default NavBar;
