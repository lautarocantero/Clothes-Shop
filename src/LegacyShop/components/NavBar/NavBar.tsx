import {
  AppBar,
  GridLegacy as Grid,
  IconButton,
  Link,
  Menu,
  MenuItem,
  Toolbar,
  Typography
} from "@mui/material";
import PropTypes from 'prop-types';
import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link as LinkReactRouter, useNavigate } from "react-router-dom";
import { startLogout, type RootState as AuthRootState } from "../../../store/auth";
import type { AppBarColor, NavBarLink } from "./types";

interface NavBarProps {
  links: NavBarLink[];
  color?: AppBarColor;
}

const NavBar = ({ links = [], color = 'primary' }: NavBarProps) => {

  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const [mobileMenuOpen, setMobileMenuOpen] = useState(false);
  const handleNavigate = useNavigate();

  const { status, name } = useSelector((state: AuthRootState) => state?.auth);

  const dispatch = useDispatch();

  const open = Boolean(anchorEl);

  const handleMouseEnter = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMouseLeave = () => {
    setAnchorEl(null);
  };

  const onLogout = () => {
    dispatch(startLogout() as any);
    handleNavigate('/auth/login');
  };

  const onRegister = () => {
    dispatch(startLogout() as any);
    handleNavigate('/auth/register');
  }

  const onHandleShoppingCart = () => {
    handleNavigate('/mi-carrito')
  }

  const DesktopMenuLinks = ({ links }: { links: NavBarLink[] }) => {
    return (
      <>
        {links?.map((linkObject: NavBarLink) => {
          const linkTo = linkObject?.linkUrl?.startsWith('/')
            ? linkObject.linkUrl
            : `/${linkObject.linkUrl}`;

          return (
            <Grid
              item
              key={crypto.randomUUID()}
              sx={{ margin: { xs: '10px', sm: '0 10px' } }}
            >
              <Link
                component={LinkReactRouter}
                to={linkTo}
                sx={{
                  fontSize: theme => theme?.typography?.body1?.fontSize,
                  textDecoration: 'none',
                  width: '100%',
                  fontWeight: 200,
                  color: theme => theme.palette?.secondary?.main,
                  '&:hover': {
                    textDecoration: 'underline',
                    color: theme => theme.palette?.secondary?.main,
                  },
                }}
              >
                {linkObject?.linkText}
              </Link>
            </Grid>
          );
        })}
      </>
    );
  };

  const FloatingMenuLinks = () => {
    if (status === 'authenticated') {
      return (
        <>
          <MenuItem onClick={onHandleShoppingCart}>My cart</MenuItem>
          <MenuItem onClick={onLogout}>Sign out</MenuItem>
        </>
      );
    }

    return (
      <>
        <MenuItem onClick={onRegister}>Sign up</MenuItem>
        <MenuItem onClick={() => handleNavigate('/auth/login')}>Login</MenuItem>
      </>
    );
  };

  const FloatingMenu = () => {
    return (
      <Menu
        open={open}
        onClose={handleMouseLeave}
        MenuListProps={{
          onMouseLeave: handleMouseLeave,
        }}
        sx={{ marginTop: '2.5em'}}
        slotProps={{
        paper: {
          sx: {
            backgroundColor: theme => theme.palette.primary.main,
            color: theme => theme.custom.white,
            borderRadius: 0,
            boxShadow: 'none',
          },
        },
        }}
        anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      >
          <FloatingMenuLinks />
        </Menu>
    )
  }

const DesktopMenu = () => {
  return (
    <>
      <Grid
        container
        alignItems="center"
        justifyContent="space-between"
        sx={{
          flexGrow: 1,
          display: { xs: 'none', sm: 'flex' },
        }}
      >
        <Grid xs={4} />
        <Grid
          item
          xs={4}
          container
          justifyContent="center"
          alignItems="center"
        >
          <DesktopMenuLinks links={links} />
        </Grid>

        <Grid
          item
          xs={4}
          container
          justifyContent="flex-end"
          alignItems="center"
        >
          <Typography sx={{ color: theme => theme?.palette?.secondary?.main, maxWidth: '200px',overflow: 'hidden',whiteSpace: 'nowrap',textOverflow: 'ellipsis'}}>{name || ''}</Typography>
          <IconButton
            onMouseEnter={handleMouseEnter}
            sx={{
              colour: theme => theme?.custom?.white,
              fontSize: theme => theme?.typography?.h4?.fontSize,
            }}
          >
            <Typography sx={{color: theme => theme?.custom?.white}}>
              <span className="material-symbols-outlined">account_circle</span>
            </Typography>
          </IconButton>
        </Grid>
      </Grid>
      <FloatingMenu />
    </>
  );
};

  const MobileMenu = () => {
    return (
      <Menu
        anchorPosition={{ top: 50, left: 0 }}
        anchorReference="anchorPosition"
        anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
        open={mobileMenuOpen}
        onClose={() => setMobileMenuOpen(false)}
        sx={{ display: { xs: 'block', sm: 'none' } }}
        slotProps={{
          paper: {
            sx: {
              width: '80%',
              backgroundColor: theme => theme.palette.primary.main,
              color: theme => theme.custom.white,
              borderRadius: 0,
              boxShadow: 'none',
            },
          },
        }}
      >
        {links?.map((linkObject: NavBarLink) => {
          const linkTo = linkObject?.linkUrl?.startsWith('/')
            ? linkObject.linkUrl
            : `/${linkObject.linkUrl}`;

          return (
            <MenuItem
              key={crypto.randomUUID()}
              component={LinkReactRouter}
              to={linkTo}
              onClick={() => setMobileMenuOpen(false)}
              sx={{ fontSize: theme => theme?.typography?.h5?.fontSize }}
            >
              {linkObject?.linkText}
            </MenuItem>
          );
        })}
        {
          (status === 'authenticated') ? 
            (<MenuItem onClick={onLogout}>Cerrar sesión</MenuItem>) : 
            (<MenuItem onClick={() => handleNavigate('/auth/login')}>Iniciar sesión</MenuItem>)
        }
      </Menu>
    );
  };

  return (
    <>
      <AppBar color={color}>
        <Toolbar>
          <Grid
            container
            sx={{
              flexDirection: { xs: 'column', sm: 'row' },
              alignContent: { xs: 'start', sm: 'none' },
            }}
          >
            <Grid>
              <IconButton
                aria-label="Abrir menú de usuario"
                sx={{ display: { xs: 'block', sm: 'none' }, color: theme => theme?.custom?.white }}
                onClick={() => setMobileMenuOpen(true)}
              >
                <span className="material-symbols-outlined">menu</span>
              </IconButton>
            </Grid>
            <DesktopMenu />
          </Grid>
        </Toolbar>
      </AppBar>
      <MobileMenu />
    </>
  );
};

NavBar.propTypes = {
  links: PropTypes.array.isRequired,
  color: PropTypes.string,
};

export default NavBar;
