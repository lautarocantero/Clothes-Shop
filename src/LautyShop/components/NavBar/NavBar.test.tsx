import {render, screen} from '@testing-library/react';
import NavBar from './NavBar';
import { MemoryRouter } from 'react-router-dom';

const linkArray = [
  { linkText: 'Tienda', linkUrl: '/'},
  { linkText: 'CÃ³mo comprar', linkUrl: '/'},
  { linkText: 'Carrito', linkUrl: '/'},
  { linkText: 'Contacto', linkUrl: '/'}
]

describe('test in NavBar Component', () => {

    test('it should match the snapshot', () => {
        const {container} = render(
            <MemoryRouter>
                <NavBar  links={linkArray} /> 
            </MemoryRouter>
        )
        expect(container).toMatchSnapshot();
    })

})