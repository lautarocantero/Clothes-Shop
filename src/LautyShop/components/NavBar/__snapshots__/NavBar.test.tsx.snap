// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test in NavBar Component > it should match the snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-26d2de-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1d159sf-MuiSvgIcon-root"
          data-testid="MenuOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiGrid-root MuiGrid-container d-flex css-ucr93j-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-1ooslc8-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1tjhsu0-MuiTypography-root-MuiLink-root"
            href="//"
          >
            Tienda
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-1ooslc8-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1tjhsu0-MuiTypography-root-MuiLink-root"
            href="//"
          >
            Cómo comprar
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-1ooslc8-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1tjhsu0-MuiTypography-root-MuiLink-root"
            href="//"
          >
            Carrito
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-1ooslc8-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1tjhsu0-MuiTypography-root-MuiLink-root"
            href="//"
          >
            Contacto
          </a>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`test in NavBar Component it should match the snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-26d2de-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1d159sf-MuiSvgIcon-root"
          data-testid="MenuOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiGrid-root MuiGrid-container d-flex css-ucr93j-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-1ooslc8-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1tjhsu0-MuiTypography-root-MuiLink-root"
            href="//"
          >
            Tienda
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-1ooslc8-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1tjhsu0-MuiTypography-root-MuiLink-root"
            href="//"
          >
            Cómo comprar
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-1ooslc8-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1tjhsu0-MuiTypography-root-MuiLink-root"
            href="//"
          >
            Carrito
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-1ooslc8-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1tjhsu0-MuiTypography-root-MuiLink-root"
            href="//"
          >
            Contacto
          </a>
        </div>
      </div>
    </div>
  </header>
</div>
`;
